types.cc

```diff
@@ -1822,6 +1822,29 @@ static void serialize(const abstract_type& t, const void* value, bytes::iterator
     return ::serialize(t, value, out, cql_serialization_format::internal());
 }
 
+template <FragmentedView View>
+data_value collection_type_impl::deserialize(View v, cql_serialization_format sf) const {
+    struct visitor {
+        View v;
+        cql_serialization_format sf;
+        data_value operator()(const abstract_type&) { abort(); }
+        data_value operator()(const list_type_impl& t) {
+            return t.deserialize(v, sf);
+        }
+        data_value operator()(const map_type_impl& t) {
+            return t.deserialize(v, sf);
+        }
+        data_value operator()(const set_type_impl& t) {
+            return t.deserialize(v, sf);
+        }
```

Comment:
Why do you need these three special cases? Can't you write `[sf, v] (const auto& t) { return t.deserialize(v, sf); }`?

Looks like the only thing the visitor does apart from that is to abort on a bad type.

Reply:
Collections have a two-argument version of deserialize (`collection_type_impl::deserialize`) which allows specifying a `cql_serialization_format`. This used to be a virtual method, but since we want to template it, we need to make it non-virtual and dispatch the call manually, with an abstract_type visitor. Just like we do in `abstract_type::deserialize`. (We could also write a `switch` over the types ourselves. Maybe we could help the compiler with optimization that way). 

You can't `[sf, v] (const auto& t) { return t.deserialize(v, sf); }` because a visitor needs to match all types, and non-collection types don't have a `deserialize` with two arguments.

Reply:
I see. Maybe they should have the two-argument form. But even if we agree they should, let's leave it for later.